File: nashorn/src/org/jetbrains/dukat/nashorn/Set.java
package org.jetbrains.dukat.nashorn;

import java.util.HashSet;

public class Set<T> {
    public int size = 0;

    private HashSet<T> hashSet = new HashSet<>();

    public Set() {
    }

    public Set(T[] elements) {
        for (T element : elements) {
            hashSet.add(element);
        }
    }

    public Set<T> add(T value) {
        hashSet.add(value);
        return this;
    }

    public void clear() {
        hashSet.clear();
    }

    public boolean delete(T value) {
        return hashSet.remove(value);
    }

    public boolean has(T value) {
        return hashSet.contains(value);
    }

    @FunctionalInterface
    public interface SetForeachCallBack<D> {
        void apply(D value, D value2, Set<D> set);
    }

    public void forEach(SetForeachCallBack<T> callBack) {
        for (T value : hashSet) {
            callBack.apply(value, value, this);
        }
    }
}

File: nashorn/src/org/jetbrains/dukat/nashorn/Set.kt
package org.jetbrains.dukat.nashorn

import java.util.HashSet

class Set<T> {
    private val hashSet = HashSet<T>()

    val size
        get(): Int {
            return hashSet.size
        }

    constructor() {}

    constructor(elements: Array<T>) {
        for (element in elements) {
            hashSet.add(element)
        }
    }

    fun add(value: T): Set<T> {
        hashSet.add(value)
        return this
    }

    fun clear() {
        hashSet.clear()
    }

    fun delete(value: T): Boolean {
        return hashSet.remove(value)
    }

    fun has(value: T): Boolean {
        return hashSet.contains(value)
    }

    fun forEach(callBack: (value: T, value2: T, set: Set<T>) -> Unit) {
        for (value in hashSet) {
            callBack(value, value, this)
        }
    }
}